### PR TITLE
Modal 2.0.0

### PR DESCRIPTION
**Pull request by on47sky**
Rethink the modal system:
- Add ability to open modals by global method, without deps on SomeModal component
- Add ability to render multiple modals with the same name

Also fix package deps.

### CODE REVIEW

File: src/standaloneModal.tsx [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  let lastId = 1
+  let activeIds: number[] = []
+
+  const Wrapper: React.FC = () => {
+    const [ , forceUpdate ] = useState(0)
+    const instanceIdRef = useRef<number>(0)
+
+    if (!instanceIdRef.current) {
+      instanceIdRef.current = lastId++
+    }
```
Comment (CleanBread) [inline src/standaloneModal.tsx]: why do we need this condition?
why can't we `const instanceIdRef = useRef<number>(1)` and without condition?

File: src/standaloneModal.tsx [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        nodes.push(
+          <Component
+            key={id}
+            name={name}
+            closeModal={closeModal}
+            {...props}
+          />
+        )
+      }
+    }
```
Comment (CleanBread) [inline src/standaloneModal.tsx:50]: why don't we need render only isStandalone modals?

File: src/standaloneModal.tsx (hunk 1 above)
Comment (on47sky) [reply src/standaloneModal.tsx]: it's an closure, each new modal will have own `instanceIdRef` value, not `1`
I thought it's more readable, but yes, we can remove the condition:
https://github.com/Local-Modules/modal/pull/1/commits/c90c53fd17c732b19c20cddb49762600c057a2d0

Comment (CleanBread) [reply src/standaloneModal.tsx]: oh, got it, thanks

File: readme.md [hunk 3]
```diff
@@ -7,199 +7,290 @@
...
-      secondaryButton: { 
-        title: 'Edit address', 
-        onClick: () => closeModal(), 
-      },
-    })
-  }, [])
+  // or temporary in useEffect
+  useEffect(() => {
+    return registerModals(modalRegistry)
+  })
```
Comment (pavelivanov) [inline readme.md]: `, [])`

Comment (pavelivanov) [reply readme.md]: what the difference of doing register in modal scope or useEffect?

File: readme.md [hunk 4]
```diff
@@ -7,199 +7,290 @@
...
-      <CommonModal />
+      <Head />
+      <Content />
+      {/* or via registry prop, which do the same in useEffect */}
+      <ModalsRenderer registry={modalRegistry} />
     </>
   )
 }
 ```
+You shouldn't add modals wrapped by `standaloneModal` to registry, use standaloneModal for not-global modals that used in specific place
```
Comment (pavelivanov) [inline readme.md]: I didn't catch how to use `standaloneModal` :)

File: readme.md [hunk 5]
```diff
@@ -7,199 +7,290 @@
...
+  // dumb but works
+  interface ModalsRegistry {
+    newModal: { priority?: number }
+  }
+}
+```
 
-import { Modal } from 'components/feedback'
-import type { ModalProps } from 'components/feedback'
+To generate properties from the component, use `ExendModalsRegistry` helper. It automatically extracts all properties from the component and removes `ModalComponentProps` from them:
```
Comment (pavelivanov) [inline readme.md]: typo Extend

File: readme.md [hunk 6]
```diff
@@ -7,199 +7,290 @@
...
+}
+```
 
-export type CustomModalProps = {
-  username: string
+## Example of global modal
+
+#### Modal Component
+```tsx
+import { type ModalComponentProps } from '@locmod/modal2'
```
Comment (pavelivanov) [inline readme.md]: @locmod/modal?

File: readme.md [hunk 7]
```diff
@@ -7,199 +7,290 @@
...
-    closeModal={closeModal}
-  >
-    <span>Hello {userName}!</span>
-  </Modal>
-)
+const AlertModal: React.FC<ModalComponentProps & AlertModalProps> = (props) => {
+  const { closeModal, title, text } = props
+  
+  const handleClick = () => {
+    // "true" will trigger onClose handler if it will be passed to props
```
Comment (pavelivanov) [inline readme.md]: I think this should be moved to separate section with detailed explanation that user can pass "onClose" prop but should not use it in the modal

File: readme.md [hunk 8]
```diff
@@ -7,199 +7,290 @@
...
+    closeModal(true)
+  }
 
-export const openCustomModal = (props: CustomModalProps) => openModal('customModal', props)
+  return (
+    <SomePlainModalMarkUpYouHave closeModal={closeModal}>
```
Comment (pavelivanov) [inline readme.md]: Also need to note somewhere that closeModal must be passed to ... wait? What is the SomePlainModalMarkUpYouHave here and why we pass closeModal there?

File: readme.md [hunk 9]
```diff
@@ -7,199 +7,290 @@
...
 }
 ```
 
-**All properties passed to `openModal(properties)` will be the props of `<CustomModal />`**
+## Example of standalone modal
 
-If you need more customization use `PlainModal`. This component provides only modal frame with overlay and close button.
+#### Modal Component
+```tsx
+import { type ModalComponentProps } from '@locmod/modal2'
```
Comment (pavelivanov) [inline readme.md]: modal?

File: src/standaloneModal.tsx (hunk 2 above)
Comment (on47sky) [reply src/standaloneModal.tsx:50]: All the system still built on unique modal names.
standaloneModal keep a modal Component and it's name in closure, it knows that it's standalone by name and verifies the name of opened modal in the state.
ModalRenderer knows nothing about modals and renders all the state (that keep all modals, global and standalone) by picking Component from the registry, that's why ModalRenderer should check `isStandalone` and skip them.

File: readme.md [hunk 10]
```diff
@@ -7,199 +7,290 @@
...
 
   return (
     <>
-      <button onClick={handleButtonClick}>Open modal</button>
-      <CustomModal />
+      <button type="button" onClick={handleButtonClick}>
+        Click me
+      </button>
+      {/* standalone modal should be rendered in place of usage */}
+      <SpecificModal />
```
Comment (pavelivanov) [inline readme.md]: The examples look too long, do we want to move them to codesandbox?

File: readme.md (hunk 3 above)
Comment (on47sky) [reply readme.md]: https://github.com/Local-Modules/modal/pull/1/commits/1a624c3ee4ec7dbf41bf343668eeb14ea449d8d5
there is no difference, except the thing that it returns an unregister function, which can be used by someone

File: readme.md (hunk 9 above)
Comment (on47sky) [reply readme.md]: https://github.com/Local-Modules/modal/pull/1/commits/3d64a29b9164e0ce30bb7d31d77959ca3dde7c46

File: readme.md (hunk 6 above)
Comment (on47sky) [reply readme.md]: https://github.com/Local-Modules/modal/pull/1/commits/3d64a29b9164e0ce30bb7d31d77959ca3dde7c46

File: readme.md (hunk 5 above)
Comment (on47sky) [reply readme.md]: https://github.com/Local-Modules/modal/pull/1/commits/3d64a29b9164e0ce30bb7d31d77959ca3dde7c46

File: readme.md (hunk 7 above)
Comment (on47sky) [reply readme.md]: https://github.com/Local-Modules/modal/pull/1/commits/fd47bf9356542beb6ed44524d388d90fec6daa4c

File: readme.md (hunk 10 above)
Comment (on47sky) [reply readme.md]: great idea
https://github.com/Local-Modules/modal/pull/1/commits/fb56bc43948815bd177e09ee84659842de44d946

File: readme.md (hunk 8 above)
Comment (on47sky) [reply readme.md]: used Dialog from headlessui in examples

File: readme.md (hunk 4 above)
Comment (on47sky) [reply readme.md]: :harold:

**Review by pavelivanov — APPROVED** [review]

**Review by CleanBread — APPROVED** [review]

**Pull request merged**